### PR TITLE
Minor bugfixes

### PR DESCRIPTION
**Pull request by eckter**
Minor bugfixes:

* Fix an issue with deep copies of TrainStates
* Remove a potential infinite loop with margins
* Reflect train deletion in debug viewer
* Fix tvd free when trains disappear
* Add tests

### CODE REVIEW

**Review by flomonster — DISMISSED** [review]

File: src/main/java/fr/sncf/osrd/infra/TVDSectionPath.java [hunk 1]
```diff
@@ -40,6 +41,10 @@ public EdgeDirection getStartTrackDirection() {
         return trackSections[0].direction;
     }
 
+    public boolean contains(TrackSectionLocation location) {
+        return Arrays.stream(trackSections).anyMatch(range -> range.containsLocation(location));
```
Comment (multun) [inline src/main/java/fr/sncf/osrd/infra/TVDSectionPath.java:45]: I'm thinking there might be a corner case where the range isn't contained by any individual range, but is contained by the union of all ranges

File: src/main/java/fr/sncf/osrd/train/Train.java [hunk 2]
```diff
@@ -101,6 +103,35 @@ public void scheduleStateChange(Simulation sim) throws SimulationError {
...
+            if (beforeTrainTail)
+                continue;
+
+            var index = currentTvdSectionPath.tvdSection.index;
+            var tvdSection = sim.infraState.getTvdSectionState(index);
+
+            tvdSection.free(sim);
+        }
+
+    }
```
Comment (multun) [inline src/main/java/fr/sncf/osrd/train/Train.java]: this seems super fishy to me, we have a queue of actions points waiting for a tail interaction, why can't we just empty this queue and signal all tvdsections the train is gone?

see https://github.com/DGEXSolutions/osrd-core/blob/853d02ff907aabe3ddd003e7a3f6889f0f47e1a5/src/main/java/fr/sncf/osrd/train/TrainState.java#L43

File: src/main/java/fr/sncf/osrd/train/TrainState.java [hunk 3]
```diff
@@ -235,6 +237,11 @@ private void step(
                 1
         );
 
+        var trackLocation = location.trackSectionRanges.getFirst().getEndLocation();
```
Comment (multun) [inline src/main/java/fr/sncf/osrd/train/TrainState.java]: can you add a comment which explains why this check is in place?

**Review by multun — CHANGES_REQUESTED** [review]

File: src/main/java/fr/sncf/osrd/train/Train.java (hunk 2 above)
Comment (eckter) [reply src/main/java/fr/sncf/osrd/train/Train.java]: It's not that simple: the tvd sections are reserved up to the end of the last route, but the train may not go that far.

There's probably a better way to write it but it's not an easy problem.

**Review by multun — DISMISSED** [review]

**Review by flomonster — APPROVED** [review]

**Review by multun — APPROVED** [review]

**Pull request merged**